File: packages/core-ts/src/graph/Graph.ts
/**
 * Graph implementation in TypeScript
 * Mirrors the Rust graph structure for unified data representation
 */

import {
  NodeId,
  EdgeId,
  PortId,
  PortDirection,
  PortType,
  INode,
  IEdge,
  IEdgeProperties,
  IPort,
} from './types';

export class Node implements INode {
  id: NodeId;
  node_type: string;
  inputs: IPort[] = [];
  outputs: IPort[] = [];
  parameters: Record<string, unknown> = {};
  state: number[] = [];

  constructor(id: NodeId, node_type: string) {
    this.id = id;
    this.node_type = node_type;
  }

  addInput(port: IPort): void {
    this.inputs.push(port);
  }

  addOutput(port: IPort): void {
    this.outputs.push(port);
  }

  setParameter(name: string, value: unknown): void {
    this.parameters[name] = value;
  }

  getParameter(name: string): unknown {
    return this.parameters[name];
  }
}

export class Port implements IPort {
  id: PortId;
  name: string;
  direction: PortDirection;
  port_type: PortType;
  value: number[] = [0];
  unit?: string;

  constructor(
    name: string,
    direction: PortDirection,
    port_type: PortType,
  ) {
    this.id = PortId.new();
    this.name = name;
    this.direction = direction;
    this.port_type = port_type;
  }

  static input(name: string, port_type: PortType): Port {
    return new Port(name, PortDirection.Input, port_type);
  }

  static output(name: string, port_type: PortType): Port {
    return new Port(name, PortDirection.Output, port_type);
  }

  setValue(value: number[]): void {
    this.value = value;
  }

  getValue(): number[] {
    return this.value;
  }

  withUnit(unit: string): Port {
    this.unit = unit;
    return this;
  }
}

export class Edge implements IEdge {
  id: EdgeId;
  source: [NodeId, PortId];
  target: [NodeId, PortId];
  properties: IEdgeProperties = {};

  constructor(source: [NodeId, PortId], target: [NodeId, PortId]) {
    this.id = EdgeId.new();
    this.source = source;
    this.target = target;
  }
}

/** Main graph data structure */
export class Graph {
  private nodes: Map<string, Node> = new Map();
  private edges: Map<string, Edge> = new Map();
  private adjacency: Map<string, string[]> = new Map();

  /** Add a node to the graph */
  addNode(node: Node): NodeId {
    const key = node.id.toString();
    this.nodes.set(key, node);
    this.adjacency.set(key, []);
    return node.id;
  }

  /** Get a node by ID */
  getNode(id: NodeId): Node | undefined {
    return this.nodes.get(id.toString());
  }

  /** Remove a node and all connected edges */
  removeNode(id: NodeId): void {
    const key = id.toString();
    if (!this.nodes.has(key)) {
      throw new Error(`Node not found: ${key}`);
    }

    // Remove connected edges
    if (this.adjacency.has(key)) {
      const edgeIds = this.adjacency.get(key) || [];
      for (const edgeId of edgeIds) {
        this.edges.delete(edgeId);
      }
    }

    // Clean up adjacency lists
    for (const edges of this.adjacency.values()) {
      const idx = edges.findIndex((eid) => {
        const edge = this.edges.get(eid);
        return edge && (edge.source[0].equals(id) || edge.target[0].equals(id));
      });
      if (idx !== -1) {
        edges.splice(idx, 1);
      }
    }

    this.nodes.delete(key);
    this.adjacency.delete(key);
  }

  /** Add an edge to the graph */
  addEdge(edge: Edge): EdgeId {
    const sourceKey = edge.source[0].toString();
    const targetKey = edge.target[0].toString();

    if (!this.nodes.has(sourceKey)) {
      throw new Error(`Source node not found: ${sourceKey}`);
    }
    if (!this.nodes.has(targetKey)) {
      throw new Error(`Target node not found: ${targetKey}`);
    }

    const edgeKey = edge.id.toString();
    this.edges.set(edgeKey, edge);

    // Update adjacency
    const sourceAdj = this.adjacency.get(sourceKey) || [];
    sourceAdj.push(edgeKey);
    this.adjacency.set(sourceKey, sourceAdj);

    const targetAdj = this.adjacency.get(targetKey) || [];
    targetAdj.push(edgeKey);
    this.adjacency.set(targetKey, targetAdj);

    return edge.id;
  }

  /** Get an edge by ID */
  getEdge(id: EdgeId): Edge | undefined {
    return this.edges.get(id.toString());
  }

  /** Remove an edge */
  removeEdge(id: EdgeId): void {
    const key = id.toString();
    const edge = this.edges.get(key);

    if (!edge) {
      throw new Error(`Edge not found: ${key}`);
    }

    this.edges.delete(key);

    // Update adjacency
    const sourceKey = edge.source[0].toString();
    const targetKey = edge.target[0].toString();

    const sourceAdj = this.adjacency.get(sourceKey);
    if (sourceAdj) {
      const idx = sourceAdj.indexOf(key);
      if (idx !== -1) {
        sourceAdj.splice(idx, 1);
      }
    }

    const targetAdj = this.adjacency.get(targetKey);
    if (targetAdj) {
      const idx = targetAdj.indexOf(key);
      if (idx !== -1) {
        targetAdj.splice(idx, 1);
      }
    }
  }

  /** Get all nodes */
  getNodes(): Node[] {
    return Array.from(this.nodes.values());
  }

  /** Get all edges */
  getEdges(): Edge[] {
    return Array.from(this.edges.values());
  }

  /** Get node count */
  nodeCount(): number {
    return this.nodes.size;
  }

  /** Get edge count */
  edgeCount(): number {
    return this.edges.size;
  }

  /** Find cycles in the graph */
  findCycles(): NodeId[][] {
    const cycles: NodeId[][] = [];
    const visited = new Set<string>();
    const recStack = new Set<string>();
    const path: NodeId[] = [];

    const dfs = (nodeId: NodeId) => {
      const key = nodeId.toString();
      visited.add(key);
      recStack.add(key);
      path.push(nodeId);

      const edgeIds = this.adjacency.get(key) || [];
      for (const edgeId of edgeIds) {
        const edge = this.edges.get(edgeId);
        if (!edge) continue;

        const nextId = edge.source[0].equals(nodeId) ? edge.target[0] : edge.source[0];
        const nextKey = nextId.toString();

        if (!visited.has(nextKey)) {
          dfs(nextId);
        } else if (recStack.has(nextKey)) {
          // Found cycle
          const idx = path.findIndex((n) => n.equals(nextId));
          if (idx !== -1) {
            cycles.push(path.slice(idx));
          }
        }
      }

      path.pop();
      recStack.delete(key);
    };

    for (const node of this.nodes.values()) {
      if (!visited.has(node.id.toString())) {
        dfs(node.id);
      }
    }

    return cycles;
  }

  /** Topological sort (returns error if cycle exists) */
  topologicalSort(): NodeId[] {
    if (this.findCycles().length > 0) {
      throw new Error('Cycle detected in graph');
    }

    const sorted: NodeId[] = [];
    const visited = new Set<string>();

    const dfs = (nodeId: NodeId) => {
      const key = nodeId.toString();
      visited.add(key);

      const edgeIds = this.adjacency.get(key) || [];
      for (const edgeId of edgeIds) {
        const edge = this.edges.get(edgeId);
        if (!edge) continue;

        const nextId = edge.source[0].equals(nodeId) ? edge.target[0] : edge.source[0];
        const nextKey = nextId.toString();

        if (!visited.has(nextKey)) {
          dfs(nextId);
        }
      }

      sorted.push(nodeId);
    };

    for (const node of this.nodes.values()) {
      if (!visited.has(node.id.toString())) {
        dfs(node.id);
      }
    }

    sorted.reverse();
    return sorted;
  }

  /** Serialize to JSON */
  toJSON(): string {
    return JSON.stringify({
      nodes: this.getNodes(),
      edges: this.getEdges(),
    });
  }

  /** Deserialize from JSON */
  static fromJSON(json: string): Graph {
    const graph = new Graph();
    const data = JSON.parse(json);

    // Deserialize nodes
    for (const nodeData of data.nodes) {
      const node = new Node(
        new NodeId(nodeData.id.value),
        nodeData.node_type,
      );
      node.parameters = nodeData.parameters;
      node.state = nodeData.state;
      graph.addNode(node);
    }

    // Deserialize edges
    for (const edgeData of data.edges) {
      const edge = new Edge(
        [new NodeId(edgeData.source[0].value), new PortId(edgeData.source[1].value)],
        [new NodeId(edgeData.target[0].value), new PortId(edgeData.target[1].value)],
      );
      edge.id = new EdgeId(edgeData.id.value);
      edge.properties = edgeData.properties;
      graph.addEdge(edge);
    }

    return graph;
  }
}

File: packages/core-rust/src/graph/graph.rs
//! Core graph implementation

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use super::node::{Node, NodeId, NodeData};
use super::edge::{Edge, EdgeData, EdgeId};
use super::port::PortId;
use crate::error::{Result, TupanError};

/// Main graph data structure that manages nodes and edges
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Graph {
    /// All nodes in the graph, indexed by ID
    nodes: HashMap<NodeId, NodeData>,
    /// All edges in the graph, indexed by ID
    edges: HashMap<EdgeId, EdgeData>,
    /// Adjacency information: node -> connected edges
    adjacency: HashMap<NodeId, Vec<EdgeId>>,
}

impl Graph {
    /// Create a new empty graph
    pub fn new() -> Self {
        Graph {
            nodes: HashMap::new(),
            edges: HashMap::new(),
            adjacency: HashMap::new(),
        }
    }

    /// Add a node to the graph
    pub fn add_node(&mut self, node: NodeData) -> NodeId {
        let id = node.id;
        self.nodes.insert(id, node);
        self.adjacency.insert(id, Vec::new());
        id
    }

    /// Get a node by ID
    pub fn get_node(&self, id: NodeId) -> Option<&NodeData> {
        self.nodes.get(&id)
    }

    /// Get a mutable reference to a node
    pub fn get_node_mut(&mut self, id: NodeId) -> Option<&mut NodeData> {
        self.nodes.get_mut(&id)
    }

    /// Remove a node and all its connected edges
    pub fn remove_node(&mut self, id: NodeId) -> Result<()> {
        if !self.nodes.contains_key(&id) {
            return Err(TupanError::NodeNotFound(id.to_string()));
        }

        // Remove all edges connected to this node
        if let Some(edge_ids) = self.adjacency.remove(&id) {
            for edge_id in edge_ids {
                self.edges.remove(&edge_id);
            }
        }

        // Remove from all adjacency lists
        for edges in self.adjacency.values_mut() {
            edges.retain(|&eid| {
                if let Some(edge) = self.edges.get(&eid) {
                    edge.source.0 != id && edge.target.0 != id
                } else {
                    false
                }
            });
        }

        self.nodes.remove(&id);
        Ok(())
    }

    /// Add an edge to the graph
    pub fn add_edge(&mut self, edge: EdgeData) -> Result<EdgeId> {
        let source_id = edge.source.0;
        let target_id = edge.target.0;
        let edge_id = edge.id;

        // Verify both nodes exist
        if !self.nodes.contains_key(&source_id) {
            return Err(TupanError::NodeNotFound(source_id.to_string()));
        }
        if !self.nodes.contains_key(&target_id) {
            return Err(TupanError::NodeNotFound(target_id.to_string()));
        }

        self.edges.insert(edge_id, edge);

        // Update adjacency
        self.adjacency.entry(source_id).or_insert_with(Vec::new).push(edge_id);
        self.adjacency.entry(target_id).or_insert_with(Vec::new).push(edge_id);

        Ok(edge_id)
    }

    /// Get an edge by ID
    pub fn get_edge(&self, id: EdgeId) -> Option<&EdgeData> {
        self.edges.get(&id)
    }

    /// Get a mutable reference to an edge
    pub fn get_edge_mut(&mut self, id: EdgeId) -> Option<&mut EdgeData> {
        self.edges.get_mut(&id)
    }

    /// Remove an edge
    pub fn remove_edge(&mut self, id: EdgeId) -> Result<()> {
        if !self.edges.contains_key(&id) {
            return Err(TupanError::EdgeNotFound(id.to_string()));
        }

        let edge = self.edges.remove(&id).unwrap();

        // Update adjacency
        let source_id = edge.source.0;
        let target_id = edge.target.0;

        if let Some(edges) = self.adjacency.get_mut(&source_id) {
            edges.retain(|&eid| eid != id);
        }
        if let Some(edges) = self.adjacency.get_mut(&target_id) {
            edges.retain(|&eid| eid != id);
        }

        Ok(())
    }

    /// Get all nodes
    pub fn nodes(&self) -> impl Iterator<Item = &NodeData> {
        self.nodes.values()
    }

    /// Get all edges
    pub fn edges(&self) -> impl Iterator<Item = &EdgeData> {
        self.edges.values()
    }

    /// Get number of nodes
    pub fn node_count(&self) -> usize {
        self.nodes.len()
    }

    /// Get number of edges
    pub fn edge_count(&self) -> usize {
        self.edges.len()
    }

    /// Find cycles in the graph (if any)
    pub fn find_cycles(&self) -> Vec<Vec<NodeId>> {
        let mut cycles = Vec::new();
        let mut visited = std::collections::HashSet::new();
        let mut rec_stack = std::collections::HashSet::new();

        for &node_id in self.nodes.keys() {
            if !visited.contains(&node_id) {
                self._dfs_cycle(&node_id, &mut visited, &mut rec_stack, &mut Vec::new(), &mut cycles);
            }
        }

        cycles
    }

    fn _dfs_cycle(
        &self,
        node_id: &NodeId,
        visited: &mut std::collections::HashSet<NodeId>,
        rec_stack: &mut std::collections::HashSet<NodeId>,
        path: &mut Vec<NodeId>,
        cycles: &mut Vec<Vec<NodeId>>,
    ) {
        visited.insert(*node_id);
        rec_stack.insert(*node_id);
        path.push(*node_id);

        if let Some(edge_ids) = self.adjacency.get(node_id) {
            for &edge_id in edge_ids {
                if let Some(edge) = self.edges.get(&edge_id) {
                    let next_id = if edge.source.0 == *node_id {
                        edge.target.0
                    } else {
                        edge.source.0
                    };

                    if !visited.contains(&next_id) {
                        self._dfs_cycle(&next_id, visited, rec_stack, path, cycles);
                    } else if rec_stack.contains(&next_id) {
                        // Found a cycle
                        if let Some(pos) = path.iter().position(|&n| n == next_id) {
                            cycles.push(path[pos..].to_vec());
                        }
                    }
                }
            }
        }

        path.pop();
        rec_stack.remove(node_id);
    }

    /// Perform topological sort (returns error if cycle exists)
    pub fn topological_sort(&self) -> Result<Vec<NodeId>> {
        if !self.find_cycles().is_empty() {
            return Err(TupanError::CycleDetected);
        }

        let mut sorted = Vec::new();
        let mut visited = std::collections::HashSet::new();

        for &node_id in self.nodes.keys() {
            if !visited.contains(&node_id) {
                self._topo_dfs(&node_id, &mut visited, &mut sorted);
            }
        }

        sorted.reverse();
        Ok(sorted)
    }

    fn _topo_dfs(
        &self,
        node_id: &NodeId,
        visited: &mut std::collections::HashSet<NodeId>,
        sorted: &mut Vec<NodeId>,
    ) {
        visited.insert(*node_id);

        if let Some(edge_ids) = self.adjacency.get(node_id) {
            for &edge_id in edge_ids {
                if let Some(edge) = self.edges.get(&edge_id) {
                    let next_id = if edge.source.0 == *node_id {
                        edge.target.0
                    } else {
                        edge.source.0
                    };

                    if !visited.contains(&next_id) {
                        self._topo_dfs(&next_id, visited, sorted);
                    }
                }
            }
        }

        sorted.push(*node_id);
    }
}

impl Default for Graph {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::port::{Port, PortType};

    #[test]
    fn test_graph_creation() {
        let graph = Graph::new();
        assert_eq!(graph.node_count(), 0);
        assert_eq!(graph.edge_count(), 0);
    }

    #[test]
    fn test_add_node() {
        let mut graph = Graph::new();
        let mut node = NodeData::new(NodeId::new(), "test".to_string());
        node.add_input(Port::input("in".to_string(), PortType::Signal));

        let node_id = graph.add_node(node);
        assert_eq!(graph.node_count(), 1);
        assert!(graph.get_node(node_id).is_some());
    }

    #[test]
    fn test_add_edge() {
        let mut graph = Graph::new();
        let node1 = NodeData::new(NodeId::new(), "node1".to_string());
        let node2 = NodeData::new(NodeId::new(), "node2".to_string());

        let id1 = graph.add_node(node1);
        let id2 = graph.add_node(node2);

        let edge = EdgeData::new((id1, PortId::new()), (id2, PortId::new()));
        let result = graph.add_edge(edge);

        assert!(result.is_ok());
        assert_eq!(graph.edge_count(), 1);
    }

    #[test]
    fn test_remove_node() {
        let mut graph = Graph::new();
        let mut node = NodeData::new(NodeId::new(), "test".to_string());
        node.add_input(Port::input("in".to_string(), PortType::Signal));

        let node_id = graph.add_node(node);
        graph.remove_node(node_id).unwrap();

        assert_eq!(graph.node_count(), 0);
    }
}
